pillarbox-player/src/main/java/ch/srgssr/pillarbox/player/analytics/metrics/MetricsCollector.kt

```diff
@@ -0,0 +1,312 @@
+            player.sessionManager.getSessionFromPeriodUid(periodUid)?.let {
+                getMetricsForSession(it)?.let(this::notifyMetricsReady)
+            }
+        })
+        loadingTimes[periodUid] = newLoadingTime
+        return newLoadingTime
+    }
+
+    override fun onStallChanged(eventTime: EventTime, isStall: Boolean) {
+        if (isStall) {
+            totalStallTimeCounter.play()
+            stallCount++
+        } else {
+            totalStallTimeCounter.pause()
+        }
+    }
+
+    override fun onIsPlayingChanged(eventTime: EventTime, isPlaying: Boolean) {
+        if (isPlaying) {
+            totalPlaytimeCounter.play()
+        } else {
+            totalPlaytimeCounter.pause()
+        }
+    }
+
+    override fun onBandwidthEstimate(eventTime: EventTime, totalLoadTimeMs: Int, totalBytesLoaded: Long, bitrateEstimate: Long) {
+        bandwidth = bitrateEstimate
+    }
+
+    override fun onVideoInputFormatChanged(eventTime: EventTime, format: Format, decoderReuseEvaluation: DecoderReuseEvaluation?) {
+        videoFormat = format
+    }
+
+    override fun onVideoDisabled(eventTime: EventTime, decoderCounters: DecoderCounters) {
+        videoFormat = null
+    }
+
+    override fun onAudioInputFormatChanged(eventTime: EventTime, format: Format, decoderReuseEvaluation: DecoderReuseEvaluation?) {
+        audioFormat = format
+    }
+
+    override fun onAudioDisabled(eventTime: EventTime, decoderCounters: DecoderCounters) {
+        audioFormat = null
+    }
+
+    private fun updateStartupTimeWithState(eventTime: EventTime, state: Int) {
+        if (eventTime.timeline.isEmpty) return
+        val periodUid = eventTime.getUidOfPeriod(window)
+        val startupTimes = getOrCreateLoadingTimes(periodUid)
+        startupTimes.state = state
+    }
+
+    override fun onPlaybackStateChanged(eventTime: EventTime, state: Int) {
+        updateStartupTimeWithState(eventTime, state)
+        when (state) {
+            Player.STATE_BUFFERING -> {
+                totalBufferingTimeCounter.play()
+            }
+
+            Player.STATE_READY -> {
+                totalBufferingTimeCounter.pause()
+            }
+        }
+    }
+
+    override fun onRenderedFirstFrame(eventTime: EventTime, output: Any, renderTimeMs: Long) {
+        updateStartupTimeWithState(eventTime, player.playbackState)
+    }
+
+    override fun onAudioPositionAdvancing(eventTime: EventTime, playoutStartSystemTimeMs: Long) {
+        updateStartupTimeWithState(eventTime, player.playbackState)
+    }
+
+    override fun onEvents(player: Player, events: AnalyticsListener.Events) {
+        bufferDuration = player.totalBufferedDuration.milliseconds
+    }
+
+    override fun onLoadCompleted(eventTime: EventTime, loadEventInfo: LoadEventInfo, mediaLoadData: MediaLoadData) {
+        if (eventTime.timeline.isEmpty) return
+        val periodUid = eventTime.getUidOfPeriod(window)
+        val loadingTimes = getOrCreateLoadingTimes(periodUid)
```

Comment:
Maybe extract that in a dedicated method, so it can be reused in `updateStartupTimeWithState` too?